File: Data Structures & Algorithms/add-two-numbers/submission-1.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */

class Solution {
public:
    ListNode* addTwoNumbers(ListNode* l1, ListNode* l2) {
        int count = 1;
        int sum = 0;
        cout << "jawn";
        while (l1 != nullptr) {
            sum += count * l1->val;
            count *= 10;
            l1 = l1->next;
        }
        cout << "made it 1 \n";
        count = 1;
        while (l2 != nullptr) {
            sum += count * l2->val;
            count *= 10;
            l2 = l2->next;
        }
        if (sum == 0) {
            return new ListNode(0, nullptr);
        }
        cout << "made it 2 \n";
        int place = 10;

        ListNode *head = nullptr;
        ListNode *curr = head;
        while (sum > 0) {
            int value = sum % place;
            sum -= value;
            value = value / (place/10);
            place *= 10;
            ListNode *new_node = new ListNode(value, nullptr);
            if (!head) {
                head = new_node;
                curr = head;
            } else {
                curr->next = new_node;
                curr = curr->next;
            }
        }
        return head;


    }
};

File: Data Structures & Algorithms/reorder-linked-list/submission-1.py
# Definition for singly-linked list.
# class ListNode:
#     def __init__(self, val=0, next=None):
#         self.val = val
#         self.next = next

class Solution:
    def reorderList(self, head: Optional[ListNode]) -> None:
        #starting at the middle of the list we should reverse it
        # then stitch the two halves togehter

        slow = head
        fast = head
        p = None
        print("1")

        while slow and fast and fast.next:
            p = slow
            slow = slow.next
            fast = fast.next.next

        # i think we want the later middle one 
        if fast:
            p = slow
            slow = slow.next
        p.next = None

        print("2")

        #slow now points to middle of the list (if its a list of 4 things points to index 2 out of 3)
        # lets reverse this slow half

        p = None
        while slow:
            next_node = slow.next
            slow.next = p
            p = slow
            slow = next_node
        print("3")

        backward = p
        forward = head
        print("made it")

        while backward:
            r1 = backward.next
            r2 = forward.next

            forward.next = backward
            backward.next = r2
            backward = r1
            forward = r2
